File: backend/src/main/java/com/distributrack/service/InventoryService.java
package com.distributrack.service;

import com.distributrack.dto.request.InventoryRequest;
import com.distributrack.dto.request.StockAdjustRequest;
import com.distributrack.dto.response.InventoryResponse;
import com.distributrack.dto.response.StockMovementResponse;

import java.util.List;

public interface InventoryService {

    // Create Inventory
    InventoryResponse createInventory(InventoryRequest request);

    // Get All Inventory
    List<InventoryResponse> getAllInventory();

    // Get Inventory By Id
    InventoryResponse getInventoryById(Long id);

    // Update Inventory
    InventoryResponse updateInventory(Long id, InventoryRequest request);

    // Delete Inventory
    void deleteInventory(Long id);

    // Get Inventory By Product
    InventoryResponse getInventoryByProduct(Long productId);

    // Get Low Stock Products
    List<InventoryResponse> getLowStockProducts();

    // Get Inventory By Warehouse
    List<InventoryResponse> getInventoryByWarehouse(String warehouseLocation);

    /**
     * Apply a stock movement (IN / OUT / ADJUSTMENT). The balance can
     * never go below zero, and every change is recorded to the movement
     * history inside the same transaction.
     */
    StockMovementResponse adjustStock(Long id, StockAdjustRequest request);

    /** Full movement history for one inventory record, newest first. */
    List<StockMovementResponse> getStockMovements(Long inventoryId);
}

File: frontend/src/services/api/inventoryService.ts
import { axiosInstance } from "@/services/api/axiosInstance";
import type {
  Inventory,
  InventoryPayload,
} from "@/types/inventory.types";

export const inventoryService = {
  async getAllInventory(): Promise<Inventory[]> {
    const response =
      await axiosInstance.get<Inventory[]>("/inventory");

    return response.data;
  },

  async getInventoryById(
    id: number,
  ): Promise<Inventory> {
    const response =
      await axiosInstance.get<Inventory>(
        `/inventory/${id}`,
      );

    return response.data;
  },

  async createInventory(
    payload: InventoryPayload,
  ): Promise<Inventory> {
    const response =
      await axiosInstance.post<Inventory>(
        "/inventory",
        payload,
      );

    return response.data;
  },

  async updateInventory(
    id: number,
    payload: InventoryPayload,
  ): Promise<Inventory> {
    const response =
      await axiosInstance.put<Inventory>(
        `/inventory/${id}`,
        payload,
      );

    return response.data;
  },

  async deleteInventory(
    id: number,
  ): Promise<string> {
    const response =
      await axiosInstance.delete<string>(
        `/inventory/${id}`,
      );

    return response.data;
  },

  async getInventoryByProduct(
    productId: number,
  ): Promise<Inventory> {
    const response =
      await axiosInstance.get<Inventory>(
        `/inventory/product/${productId}`,
      );

    return response.data;
  },

  async getLowStockProducts(): Promise<Inventory[]> {
    const response =
      await axiosInstance.get<Inventory[]>(
        "/inventory/low-stock",
      );

    return response.data;
  },

  async getInventoryByWarehouse(
    warehouseLocation: string,
  ): Promise<Inventory[]> {
    const response =
      await axiosInstance.get<Inventory[]>(
        `/inventory/warehouse/${encodeURIComponent(
          warehouseLocation,
        )}`,
      );

    return response.data;
  },
};